src/libpmemfile-posix/file.c

```diff
@@ -997,6 +997,143 @@ pmemfile_unlink(PMEMfilepool *pfp, const char *pathname)
 	return pmemfile_unlinkat(pfp, PMEMFILE_AT_CWD, pathname, 0);
 }
 
+struct pmemfile_dirent_info {
+	struct pmemfile_vinode *vinode;
+	struct pmemfile_dirent *dirent;
+};
+
+/*
+ * lock_parents_and_children -- resolve 2 files with respect to parent
+ * directories and lock all 4 inodes in write mode
+ *
+ * Returns 0 on success.
+ * Returns negated errno when failed.
+ * Returns 1 when there was a race.
+ */
+static int
+lock_parents_and_children(PMEMfilepool *pfp,
+		struct pmemfile_path_info *src,
+		struct pmemfile_dirent_info *src_info,
+
+		struct pmemfile_path_info *dst,
+		struct pmemfile_dirent_info *dst_info,
+
+		struct pmemfile_vinode *vinodes[4],
+		size_t *vinodes_num)
+{
+	memset(src_info, 0, sizeof(*src_info));
+	memset(dst_info, 0, sizeof(*dst_info));
+
+	size_t src_namelen = component_length(src->remaining);
+	size_t dst_namelen = component_length(dst->remaining);
+
+	/* lock 2 parents in correct order */
+	vinode_rdlock2(src->vinode, dst->vinode);
+
+	/* find source file */
+	src_info->dirent = vinode_lookup_dirent_by_name_locked(pfp, src->vinode,
+				src->remaining, src_namelen);
+	if (!src_info->dirent) {
+		int error = errno;
+
+		vinode_unlock2(src->vinode, dst->vinode);
+
+		return -error;
+	}
+
+	/* get the vinode for found source file */
+	src_info->vinode = inode_ref(pfp, src_info->dirent->inode, src->vinode,
+					NULL, src->remaining, src_namelen);
+	if (!src_info->vinode) {
+		int error = errno;
+
+		vinode_unlock2(src->vinode, dst->vinode);
+
+		return -error;
+	}
+
+	/* find destination file (it may not exist) */
+	dst_info->dirent = vinode_lookup_dirent_by_name_locked(pfp, dst->vinode,
+					dst->remaining, dst_namelen);
+	if (dst_info->dirent) {
+		/* if file exists get the vinode for it */
+		dst_info->vinode = inode_ref(pfp, dst_info->dirent->inode,
+				dst->vinode, NULL, dst->remaining, dst_namelen);
+		if (!dst_info->vinode) {
```

Comment:
Just asking, the `inode_ref` thing can only fail, if an underlying `malloc` fails -- is that right?